### ISSUE TITLE
Any plan to bump QEMU version?

### ISSUE THREAD

**Issue by cnSchwarzer** [unicorn-engine/unicorn#1798]
I see unicorn is inevitably strongly coupled to QEMU, so it is hard to merge upstream changes since QEMU is completely off track.
However the current version 5.0.1 is kind of out-dated and some new target like RISC-V has many changes since 2019.
Is there any plan to keep the track?

**Comment by wtdcode** [comment]
Yes, we indeed plan to do so but as said in the FAQ it's impossible to always keep Unicorn sync-ed with upstream QEMU.

**Comment by aquynh** [comment]
We want to, but that is a lot of work. And our time is not paid.

**Comment by cnSchwarzer** [comment]
Thanks for your good work! I find out that you modified all tcg functions to make it instantiable, which is a huge work if not tools assisted.

**Comment by cnSchwarzer** [comment]
A probably fool question: 
Since tcg context is thread local, why not create thread for each uc instance to isolate the tcg contexts, which we can avoid modify all tcg functions with a this pointer.

**Comment by wtdcode** [comment]
> A probably fool question: Since tcg context is thread local, why not create thread for each uc instance to isolate the tcg contexts, which we can avoid modify all tcg functions with a this pointer.

Porting mtcg is not helpful for Unicorn API design I think.

**Comment by XVilka** [comment]
The current latest version of QEMU is 8.1: https://wiki.qemu.org/ChangeLog/8.1

You might consider eventual porting of at least some of the features to TCG plugins, like some other projects do:
- https://lore.kernel.org/qemu-devel/20230816145155.21049-1-akihiko.odaki@daynix.com/T/#md4f5cc5eecdd2cb5b0672d760dc811c32aa68ba3
- https://github.com/panda-re/panda/issues/570

**Comment by wtdcode** [comment]
> The current latest version of QEMU is 8.1: https://wiki.qemu.org/ChangeLog/8.1
> 
> You might want to consider eventual porting of at least some of the features to TCG plugins, like some other projects do:
> 
> * https://lore.kernel.org/qemu-devel/20230816145155.21049-1-akihiko.odaki@daynix.com/T/#md4f5cc5eecdd2cb5b0672d760dc811c32aa68ba3
> * [Update to modern QEMU! panda-re/panda#570](https://github.com/panda-re/panda/issues/570)

I have been keeping an eye on TCG plugins development and I think it's possible to migrate all unicorn-specific to plugins, though meaning bunch of rework.

**Comment by aquynh** [comment]
While there are some overlap, Unicorn is quite different from that plugin
architecture of Qemu, so that is not a solution in my view.

**Comment by daxie117** [comment]
can i request some elaboration here?

is the concern primarily the fact that TCG plugins are merely observers of system state, whereas the arch of unicorn engine allows the user to completely control and manipulate system state?

**Comment by wtdcode** [comment]
Yes exactly.


From: daxie117 ***@***.***>
Date: Friday, 10 May 2024 at 7:07 AM
To: unicorn-engine/unicorn ***@***.***>
Cc: lazymio ***@***.***>, Comment ***@***.***>
Subject: Re: [unicorn-engine/unicorn] Any plan to bump QEMU version? (Issue #1798)

can i request some elaboration here?

is the concern primarily the fact that TCG plugins are merely observers of system state, whereas the arch of unicorn engine allows the user to completely control and manipulate system state?

**Comment by wtdcode** [comment]
> QEMU developers now are working on adding TCG plugins API to also write registers and memory, not only read them. It might be exactly what Unicorn needs: https://lore.kernel.org/qemu-devel/20250523020344.1341179-1-rowanbhart@gmail.com/

Thanks for raising this. This moves a big step but still as aquynh pointed out, it is not enough to build Unicorn solely on tcg plugins.
